proto/message.py

```diff
@@ -572,65 +606,64 @@ def __eq__(self, other):
-        sequence.
-
-        For messages, protocol buffers does distinguish between an empty
-        message and absence, but this distinction is subtle and rarely
-        relevant. Therefore, this method always returns an empty message
-        (following the official implementation). To check for message
-        presence, use ``key in self`` (in other words, ``__contains__``).
-
-        .. note::
-
-            Some well-known protocol buffer types
-            (e.g. ``google.protobuf.Timestamp``) will be converted to
-            their Python equivalents. See the ``marshal`` module for
-            more details.
-        """
-        try:
-            pb_type = self._meta.fields[key].pb_type
-            pb_value = getattr(self._pb, key)
-            marshal = self._meta.marshal
-            return marshal.to_python(pb_type, pb_value, absent=key not in self)
-        except KeyError as ex:
-            raise AttributeError(str(ex))
-
     def __ne__(self, other):
         """Return True if the messages are unequal, False otherwise."""
         return not self == other
 
     def __repr__(self):
         return repr(self._pb)
 
-    def __setattr__(self, key, value):
-        """Set the value on the given field.
+    @property
+    def _pb(self):
+        if self._cached_pb is None:
+            self._cached_pb = self._meta.pb()
+        return self._cached_pb
+
+    @_pb.setter
+    def _pb(self, value):
+        self._cached_pb = value
+
+    def _mark_pb_stale(self, field_name: str):
+        """We often set fields on the proto-plus object (this), and do not immediately
+        sync them down to the underlying pb2 object. This mechanism tracks which
+        fields are stale in this way.
+        """
+        if not hasattr(self, '_stale_fields'):
+            self._stale_fields = set()
+        self._stale_fields.add(field_name)
 
-        For well-known protocol buffer types which are marshalled, either
-        the protocol buffer object or the Python equivalent is accepted.
+    def _mark_pb_synced(self):
+        self._stale_fields = set()
+
+    def _update_nested_pb(self):
+        """When it is time to serialize a pb2 object, it does not do to sync just ourselves -
+        we must also recursively search for nested proto-plus objects that may also require
+        a sync.
         """
-        if key[0] == "_":
-            return super().__setattr__(key, value)
-        marshal = self._meta.marshal
-        pb_type = self._meta.fields[key].pb_type
-        pb_value = marshal.to_proto(pb_type, value)
+        for field_name, field in self._meta.fields.items():
+            if field.proto_type == proto.MESSAGE:
+                obj = getattr(self, field_name, None)
+                if obj and hasattr(obj, '_update_pb'):
+                    obj._update_pb()
+                if obj and hasattr(obj, '_update_nested_pb'):
+                    obj._update_nested_pb()
```

Comment:
It took me a few readthroughs to understand why a field value may be a message and not have `_update_pb` or `_update_nested_pb`. I bet that if we changed the conditional logic a bit it might be more clear.
```python
if obj and isinstance(obj, Message):
    obj._update_pb()
    obj._update_nested_pb()
```
I still need to mull over this logic a little bit. It seems like we might be able to skip fields under certain circumstances. Also, I'm not sure if the ordering matters. I'll need to do some experiments and come up with examples and or counterexamples.